edge/pkg/devicetwin/dttype/types_test.go

```diff
@@ -0,0 +1,623 @@
+			edgeversion:  string(bytesTwinEdgeVersion),
+			want:         true,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if got := CompareVersion(test.cloudversion, test.edgeversion); got != test.want {
+				t.Errorf("CompareVersion() = %v, want %v", got, test.want)
+			}
+		})
+	}
+}
+
+// TestUnmarshalConnectedInfo is function to test UnmarshalConnectedInfo().
+func TestUnmarshalConnectedInfo(t *testing.T) {
+	connected := ConnectedInfo{
+		EventType: "Event",
+		TimeStamp: time.Now().UnixNano() / 1e6,
+	}
+	bytesConnected, err := json.Marshal(connected)
+	if err != nil {
+		t.Errorf("Error in Marshalling: %v", err)
+	}
+	tests := []struct {
+		name     string
+		argument []byte
+		want     ConnectedInfo
+		wantErr  error
+	}{
+		{
+			// Success Case
+			name:     "UnmarshalConnectedInfoTest-CorrectInput",
+			argument: bytesConnected,
+			want:     connected,
+			wantErr:  nil,
+		},
+		{
+			// Failure Case
+			name:     "UnmarshalConnectedInfoTest-WrongInput",
+			argument: []byte(""),
+			want:     ConnectedInfo{},
+			wantErr:  errors.New("unexpected end of JSON input"),
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got, err := UnmarshalConnectedInfo(test.argument)
+			if err != nil && err.Error() != test.wantErr.Error() {
+				t.Errorf("UnmarshalConnectedInfo() error = %v, wantErr %v", err, test.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, test.want) {
+				t.Errorf("UnmarshalConnectedInfo() = %v, want %v", got, test.want)
+			}
+		})
+	}
+}
+
+// TestUnmarshalDeviceTwinDocument is function to test UnmarshalDeviceTwinDocument().
+func TestUnmarshalDeviceTwinDocument(t *testing.T) {
+	var devTwinDoc DeviceTwinDocument
+	byteDevTwinDoc, err := json.Marshal(devTwinDoc)
+	if err != nil {
+		t.Errorf("Error in Marshalling: %v", err)
+	}
+	tests := []struct {
+		name     string
+		argument []byte
+		want     *DeviceTwinDocument
+		wantErr  error
+	}{
+		{
+			// Success Case
+			name:     "UnmarshalDeviceTwinDocumentTest-CorrectInput",
+			argument: byteDevTwinDoc,
+			wantErr:  nil,
+			want:     &devTwinDoc,
+		},
+		{
+			// Failure Case
+			name:     "UnmarshalDeviceTwinDocumentTest-WrongInput",
+			argument: []byte(""),
+			want:     &DeviceTwinDocument{},
+			wantErr:  errors.New("unexpected end of JSON input"),
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got, err := UnmarshalDeviceTwinDocument(test.argument)
+			if err != nil && err.Error() != test.wantErr.Error() {
+				t.Errorf("UnmarshalDeviceTwinDocument() error = %v, wantErr %v", err, test.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, test.want) {
+				t.Errorf("UnmarshalDeviceTwinDocument() = %v, want %v", got, test.want)
+			}
+		})
+	}
+}
+
+//TestUnmarshalDeviceTwinUpdate is function to test UnmarshalDeviceTwinUpdate().
+func TestUnmarshalDeviceTwinUpdate(t *testing.T) {
+	//Empty DeviceTwin variable
```

Comment:
please move all the data contstruction code to separate functions which could be possibly reused. the test can simply use these helper methods.

Reply:
Updated